idseq_pipeline/commands/non_host_alignment_functions.py

```diff
@@ -709,28 +633,83 @@ def run_gsnapl_chunk(part_suffix, remote_home_dir, remote_index_dir, remote_work
+
     write_to_log("finished alignment for chunk %s" % chunk_id)
-    return os.path.join(CHUNKS_RESULT_DIR, dedup_outfile_basename)
+    return multihit_local_outfile
 
 
+def call_hits_m8(input_m8, output_m8, output_summary):
+    # Get lineage info
+    lineage_path = fetch_reference(LINEAGE_SHELF)
+    lineage_map = shelve.open(lineage_path)
+    accession2taxid_path = fetch_reference(ACCESSION2TAXID)
+    accession2taxid_dict = shelve.open(accession2taxid_path)
+    # Helper functions
+    def add_taxid_hits(accession, hits):
+        taxid = accession2taxid_dict.get(accession.split(".")[0], "NA")
+        species_taxid, genus_taxid, family_taxid = lineage_map.get(taxid, ("-100", "-200", "-300"))
+        hits["species"] += [species_taxid]
+        hits["genus"] += [genus_taxid]
+        hits["family"] += [family_taxid]
+        return hits
+    def call_hit_level(read_id, hits):
+        for i, level in enumerate(["species", "genus", "family"]):
+            taxids = hits[level]
+            taxids = [t for t in taxids if int(t) >= 0] # do not consider unclassified hits. TO DO: decide if that's an improvement
+            n = len(set(taxids)) # number of distinct taxids at this level
+            if n == 1:
+                return i+1, taxids[0]
+        return -1, -1
+    # Deduplicate m8 and summarize hits
+    read_ids = subprocess.check_output("grep -v '^#' %s | cut -f1" % input_m8, shell="true").split("\n")
+    read_ids = filter(None, read_ids)
+    read_ids = set(read_ids)
+    sorted_input_m8 = input_m8 + "-sorted"
+    execute_command("sort -k1 %s > %s" % (input_m8, sorted_input_m8))
+    read_to_hit_level = {}
+    read_to_taxid = {}
+    outf = open(output_m8, "wb")
+    outf_sum = open(output_summary, "wb")
+    for read_id in read_ids:
+        # TO DO: remove inefficiency of calling grep on the entire file for each read_id.
+        # Lines with same read_id are contiguous (verify?), so just iterate line by line.
+        m8_lines = subprocess.check_output("grep '^%s\t' %s" % (read_id, sorted_input_m8), shell="true").split("\n")
+        accessions_evalues_lines = [(line.split("\t")[1],
+                                     float(line.split("\t")[10]),
+                                     line) for line in m8_lines if line]
+        best_evalue = min([item[1] for item in accessions_evalues_lines])
+        best_accessions_evalues_lines = [item for item in accessions_evalues_lines if item[1] == best_evalue]
+        best_accessions = [item[0] for item in best_accessions_evalues_lines]
+
+        first_line = best_accessions_evalues_lines[0][2]
+        outf.write(first_line + "\n")
+
+        hits = { "species": [], "genus": [], "family": [] }
+        for acc in best_accessions:
+            hits = add_taxid_hits(acc, hits)
+        hit_level, taxid = call_hit_level(read_id, hits)
+        outf_sum.write("%s\t%d\t%s\n" % (read_id, hit_level, taxid))
+    outf.close()
+    outf_sum.close()
```

Comment:
Because this is running inside multiple threads, we don't want to leak open file descriptors if exceptions occur in some of the threads, because good hygiene there would increase the chances of the unaffected chunks to complete.   The classic way of doing that is
```
    f.open()
    try:
        do some work on f
    finally:
        f.close()
```
but python has a thing for this that handles it 
```
with open("...", "") as f:
    do some work on f
```
since for two files we'll have to nest, that means deeper indentation, but that's worth it for the built in exception handling

Reply:
ok